proposals/0239-operator-designated-types.md

```diff
@@ -0,0 +1,159 @@
+let i: Int? = 1
+let j: Int?
+let k: Int? = 2
+
+let _ = [i, j, k].reduce(0 as Int?) {
+  $0 != nil && $1 != nil ? $0! + $1! : ($0 != nil ? $0! : ($1 != nil ? $1! : nil))
+}
+
+
+// Was "too complex", now 5ms:
+func test(a: [String], b: String, c: String) -> [String] {
+  return a.map { $0 + ": " + b + "(" + c + $0 + ")" }
+}
+
+// Was "too complex", now 25ms:
+func test(strings: [String]) {
+  for string in strings {
+    let _ = string.split(omittingEmptySubsequences: false) { $0 == "C" || $0 == "D" || $0 == "H" || $0 == "S"}
+  }
+}
+
+// Was "too complex", now 5ms:
+func test(n: Int) -> Int {
+  return n == 0 ? 0 : (0..<n).reduce(0) {
+    ($0 > 0 && $1 % 2 == 0) ? ((($0 + $1) - ($0 + $1)) / ($1 - $0)) + (($0 + $1) / ($1 - $0)) : $0
+  }
+}
+
+// Was "too complex", now 2ms:
+let a = 1
+_ = -a + -a - -a + -a - -a
+
+```
+
+A number (25+) of other test cases were previously compiling quickly were also tested using this functionality and with the existing hacks disabled, and those cases are as fast or faster than previously.
+
+## Source compatibility
+
+Because of the fact that we fall back to behavior that is very similar to the current behavior of the type checker (exhaustively testing combinations of operators in the cases), this change is very compatible with the current behavior. 
+
+One failure was seen in the source compatibility suite when running with this new behavior enabled by default in all language modes. The failure was in `BlueSocket`, and that same failure also manifests if you build the affected code with `-swift-version 4`. Specifically, for `let mask = 1 << (2 as Int32)`, we now infer `mask` as `Int` rather than `Int32`.
+
+It's possible to construct cases where behavior will change based on first attempting the overloads defined in the designated type and then checking no other alternatives if that overload succeeds. It's difficult to quantify how often existing code might hit this, but the same code would also potentially break due to implementation changes in the type checker. This proposal makes breaks like this less likely in the future by codifying another piece of behavior in the type checker.
+
+One example of where we can expect to see breaks from this change is where users implement operators that are declared in the standard library for types that would also typecheck successfully with the standard library implementation. For example a user has a type that is `Equatable` and has defined `==` where one or both operands are the user's type, but where the implementation behaves differently than calling the `==` defined on `Equatable`.
+
+We can see this in the [`Anchorage`](https://github.com/RaizLabs/Anchorage) project on GitHub.
+
+```
+public func == <T: BinaryFloatingPoint>(lhs: NSLayoutDimension, rhs: T) -> NSLayoutConstraint {
+    return finalize(constraint: lhs.constraint(equalToConstant: CGFloat(rhs)))
+}
+```
+
+Here, `NSLayoutDimension` and `T` are both `Equatable`, and as a result code like:
```

Comment:
Are you sure this is correct?

The homogeneous `==` (two `Self` parameters) doesn't seem to be interchangeable with the generic heterogeneous `==` (`NSLayoutDimension` and `T: BinaryFloatingPoint` parameters).

If the operands are two `.widthAnchor: NSLayoutDimension & Equatable` values, I'd have thought the alternatives are:

```swift
public protocol Equatable {
  static func == (lhs: Self, rhs: Self) -> Bool
}
```

and [this one](https://github.com/Raizlabs/Anchorage/blob/89c3b605638a2883e954cc0b7bea2c2f19daf7fd/Source/Anchorage.swift#L57-L59):

```swift
public func == (lhs: NSLayoutDimension, rhs: NSLayoutDimension) -> NSLayoutConstraint
```

Reply:
Fixed in 0c9072860781b5e28bffdbe771d3741c97df16c6.